aiida/work/processes.py

```diff
@@ -477,6 +445,47 @@ def _add_description_and_label(self):
             if label is not None:
                 self._calc.label = label
 
+    def _flat_inputs(self):
+        """
+        Return a flattened version of the parsed inputs dictionary. The eventual
+        keys will be a concatenation of the nested keys
+
+        :return: flat dictionary of parsed inputs
+        """
+        return self._flatten_inputs(self.inputs)
+
+    def _flatten_inputs(self, dictionary, parent_key='', separator='_'):
+        """
+        Function that will recursively flatten the inputs dictionary, omitting
+        inputs whose key starts with an underscore as they are considered to be
+        non storable.
+
+        :param dictionary: nested dictionary of parsed inputs
+        :param parent_key: the parent key with which to prefix the keys
+        :param separator: character to use for the concatenation of keys
+        """
+        items = []
+        for key, value in dictionary.iteritems():
+
+            prefixed_key = parent_key + separator + key if parent_key else key
+
+            if isinstance(value, collections.MutableMapping):
+                items.extend(self._flatten_inputs(value, prefixed_key, separator=separator).iteritems())
+            else:
+                try:
+                    port = self.spec().inputs[key]
+                except KeyError:
+                    # It's not in the spec, so we better support dynamic inputs
+                    assert self.spec().inputs.dynamic
+                    items.append((prefixed_key, value))
+                else:
+                    # Ignore any inputs that should not be saved
+                    if port.non_db:
+                        continue
+
+                items.append((prefixed_key, value))
```

Comment:
This will be called a second time if the port does not exist in the spec (i.e. is dynamic).